### DISCUSSION TITLE
How does a terminal emulator know when it is safe to push a line into the scrollback?

### DISCUSSION THREAD

**Discussion by shi-yan** [wezterm/wezterm#3356] [Q&A] [answered]
This is a general question regarding VT100. 

How can a terminal emulator know when to push a line into the scrollback buffer? A naive implementation would do so when seeing '\n' for example. But because VT100 allows moving the cursor. If the cursor has been moved to a previous line, the previous line can be mutated. It seems that the scrollback is never an append-only buffer. This seems to increase the difficulty of implementing a terminal emulator by a lot because a previously parsed line can be changed at any time?

How does a real terminal emulator handle this? Thanks

**Comment by wez** [comment, marked as answer]
wezterm handles this by treating whatever is in the viewport as a fully mutable grid of cells.
Escape sequences are parsed and their effects are applied to the grid.
Parsing cannot assume that the input is ever fully line based, because the applications in the terminal can send commands that move the cursor after every byte and that movement could be anywhere on the screen.

When a newline is processed, if the cursor position would move off the bottom of the screen, and the scroll margins are set to match the full height of the viewport, then the top row of the grid is moved into immutable scrollback, and a new blank row is added at the bottom, which makes the general idea of viewport -> scrollback very simple.

There are complications though. If you implement reflowing of text on resize, it can impact the cursor position and there is some potential for the new cursor location to end up in the immutable scrollback and damage it when eg: the shell prompt is later rendered.

https://github.com/wez/wezterm/blob/main/term/src/screen.rs has a lot of the logic for this in wezterm's code base.

**Reply by shi-yan** [reply]
Thank you so much for the reply and point me to the code! 

Is there a spec or document for vt100? I found it difficult to find a definition for some of the terms, like scroll margin, line feed, etc.

**Reply by wez** [reply]
Take a look at https://vt100.net/
